python/mxnet/gluon/block.py

```diff
@@ -230,6 +256,7 @@ def params(self):
     def collect_params(self):
         """Returns a :py:class:`ParameterDict` containing this :py:class:`Block` and all of its
         children's Parameters."""
+        self._check_container_with_block()
```

Comment:
This should be checked on assignment

Reply:
If we check it on assignment, we cannot catch the case:
```python
def __init__(self):
    self.blocks = []
    with self.name_scope():
        for i in range(3):
            self.blocks.append(...)
```

Reply:
@piiswrong Would it be okay to accept the current approach? The collect_params will always be called if the users want to initialize the network and they may notice the warning.

Reply:
I agree that collect_params is probably a good place to put this check, because that's where the problem of not properly registering the blocks would surface.